logstash-core/lib/logstash/pipelines_registry.rb

```diff
@@ -0,0 +1,142 @@
+# encoding: utf-8
+
+module LogStash
+  class PipelineState
+    attr_reader :pipeline_id, :pipeline, :lock
+
+    def initialize(pipeline_id, pipeline)
+      @pipeline_id = pipeline_id
+      @pipeline = pipeline
+      @terminated = Concurrent::AtomicBoolean.new(false)
+      @lock = Mutex.new
+    end
+
+    def terminated?
+      @terminated.value
+    end
+
+    def set_terminated(is_terminated)
+      @terminated.value = is_terminated
+    end
+
+    def set_pipeline(pipeline)
+      @pipeline = pipeline
+    end
+  end
+
+  class PipelinesRegistry
+    attr_reader :states
+    include LogStash::Util::Loggable
+
+    def initialize
+      @states = java.util.concurrent.ConcurrentHashMap.new
+    end
+
+    def create_pipeline(pipeline_id, pipeline, &create_block)
+      success = false
+
+      @states.compute(pipeline_id) do |_, state|
+        if state
+          state.lock.synchronize do
+            if state.terminated?
+              success = create_block.call
```

Comment:
I am not sure if JRuby has the same behaviour, it'd be interesting to measure. In this case this is nowhere near a critical path so performance is not a concern here - the reason I like this way of handling blocks is because the block parameter is explicit and helps better understand the intent of the method.